pages/book/debug.mdx

```diff
@@ -1,166 +1,578 @@
+
+* Do you know the intent of the code? It's often more difficult to debug someone else's code. If it's not your code, it's possible you might need to spend time learning exactly what the code does before you can debug it effectively.
+
+<Callout>
+
+  When writing contracts, start small and start with code that works! Sometimes, it is easier to fix a large or complicated set of code by starting with a small piece of code that demonstrates the core task you are trying to achieve. Then, you can modify or add code incrementally, testing at each point for errors.
+
+  Here, it may be helpful to test your assumptions in a small [experimental playground](#lab) before rolling out a complete implementation.
+
+</Callout>
+
+### Go over your code and observe the values [#step-3]
+
+At the moment, Tact doesn't have a step-through debugger. Despite that, it's still possible to use the [_printf debugging_](https://en.wikipedia.org/wiki/Debugging#printf_debugging) approach.
+
+It involves actively placing [`dump(){:tact}`][dump] and [`dumpStack(){:tact}`](/ref/core-debug#dumpstack) function calls throughout your code and observing states of variables at a given point of time. Note, that those functions work only in a [debug mode](#debug-mode) and won't be executed otherwise.
+
+Once you found that some value isn't equal to what you've expected it to be, don't rush to fixing the issue on the spot. That's because what you're seeing may not be the root cause of it and merely a symptom, effect. Be very careful with cause-and-effect relationships and figure out which's which to resolve the cause and not introduce new mess for your future self.
+
+In addition to dumping values, it's often helpful to use assertive functions like [`require(){:tact}`](/ref/core-debug#require), [`nativeThrowWhen(){:tact}`](/ref/core-debug#nativethrowwhen) and [`nativeThrowUnless(){:tact}`](/ref/core-debug#nativethrowunless). They help stating your assumptions clear, and are handy for setting up "trip wires" for catching issues in the future.
+
+And if you didn't find or cannot resolve the cause of your issues, try asking the community in Tact's [Telegram chat][tg] or, if your issue or question is generally related to TON more than it's related to Tact, hop into [TON Dev Telegram chat](https://t.me/tondev_eng).
+
+</Steps>
+
+## Common debugging functions [#debug-functions]
+
+Tact provides a handful amount of various functions useful for debugging: [Core library → Debug](/ref/core-debug).
+
+The most commonly used debugging function is [`dump(){:tact}`][dump], which prints the passed argument to the contract's debug console. It only works in a [debug mode](#debug-mode) and does nothing otherwise.
+
+In practice, when you write:
+
+```tact {2}
+// ... somewhere appropriate in your contract ...
+dump("To infinity and beyond!");
 ```
 
-## Using `@tact-lang/emulator`
+You'll see the following in the [test logs](#tests) whenever execution gets to that [`dump(){:tact}`][dump] call:
```

Comment:
Let's provide an in-place example here. What are exact steps for me as a Blueprint/Tact user to observe the dumped values in my contract? What do I need to execute to get those? And how exactly the test output would look like? This needs to be as clear as possible, it's one of the most frequently asked questions by the novice users

Reply:
Got it

Reply:
Provided an in-place example and placed reference links to it from earlier in the page